statsdaemon.go

```diff
@@ -312,147 +343,233 @@ func processTimers(buffer *bytes.Buffer, now int64, pctls Percentiles) int64 {
+		if cap(buf) >= end {
+			buf = buf[:end]
+		} else {
+			tmp := buf
+			buf = make([]byte, end)
+			copy(buf, tmp)
+		}
 
-		index = bytes.IndexByte(input, '|')
-		if index < 0 || index == len(input) - 1 {
-			if *debug {
-				log.Printf("ERROR: failed to parse line: %s\n", string(line))
+		n, err := mp.reader.Read(buf[idx:])
+		buf = buf[:idx+n]
+		if err != nil {
+			if err != io.EOF {
+				log.Printf("ERROR: %s", err)
 			}
-			continue
-		}
 
-		val := input[:index]
-		index++
+			mp.done = true
 
-		var mtypeStr string
+			line, rest = mp.lineFrom(buf)
+			if line != nil {
+				mp.buffer = rest
+				return parseLine(line), len(rest) > 0
+			}
 
-		if input[index] == 'm' {
-			index++
-			if index >= len(input) || input[index] != 's' {
-				if *debug {
-					log.Printf("ERROR: failed to parse line: %s\n", string(line))
-				}
-				continue
+			if len(rest) > 0 {
+				return parseLine(rest), false
 			}
-			mtypeStr = "ms"
-		} else {
-			mtypeStr = string(input[index])
+
+			return nil, false
 		}
+	}
+}
 
-		index++
-		input = input[index:]
+func (mp *MsgParser) lineFrom(input []byte) ([]byte, []byte) {
+	split := bytes.SplitAfterN(input, []byte("\n"), 2)
+	if len(split) == 2 {
+		return split[0][:len(split[0])-1], split[1]
+	}
 
-		var (
-			value interface{}
-			err   error
-		)
+	if !mp.partialReads {
+		if len(input) == 0 {
+			input = nil
+		}
+		return input, []byte{}
+	}
 
-		if mtypeStr[0] == 'c' {
-			value, err = strconv.ParseInt(string(val), 10, 64)
-			if err != nil {
-				log.Printf("ERROR: failed to ParseInt %s - %s", string(val), err)
-				continue
-			}
-		} else if mtypeStr[0] == 'g' {
-			var relative, negative bool
-			var stringToParse string
-
-			switch val[0] {
-			case '+', '-':
-				relative = true
-				negative = val[0] == '-'
-				stringToParse = string(val[1:])
-			default:
-				relative = false
-				negative = false
-				stringToParse = string(val)
-			}
+	if bytes.HasSuffix(input, []byte("\n")) {
+		return input[:len(input)-1], []byte{}
+	}
 
-			gaugeValue, err := strconv.ParseUint(stringToParse, 10, 64)
-			if err != nil {
-				log.Printf("ERROR: failed to ParseUint %s - %s", string(val), err)
-				continue
-			}
+	return nil, input
+}
 
-			value = GaugeData{relative, negative, gaugeValue}
-		} else if mtypeStr[0] == 's' {
-			value = string(val)
-		} else {
-			value, err = strconv.ParseUint(string(val), 10, 64)
+func parseLine(line []byte) *Packet {
+	split := bytes.SplitN(line, []byte{'|'}, 2)
+	if len(split) < 2 {
+		logParseFail(line)
+		return nil
+	}
+
+	keyval := split[0]
+	typeCode := string(split[1])
+
+	split = bytes.SplitN(keyval, []byte{':'}, 2)
+	if len(split) < 2 {
+		logParseFail(line)
+		return nil
+	}
+	name := string(split[0])
+	val := split[1]
+	if len(val) == 0 {
+		logParseFail(line)
+		return nil
+	}
+
+	var (
+		err      error
+		value    interface{}
+		sampling float32 = 1
+	)
+
+	if strings.HasPrefix(typeCode, "c") {
+		split = bytes.SplitN([]byte(typeCode), []byte("|@"), 2)
```

Comment:
Timers also support sampling rates, may want to generalize this.

Test wise, could add this to TestParseLineTimer:
```
d = []byte("glork:320|ms|@0.1")
    packet = parseLine(d)
    assert.NotEqual(t, packet, nil)
    assert.Equal(t, "glork", packet.Bucket)
    assert.Equal(t, uint64(320), packet.Value.(uint64))
    assert.Equal(t, "ms", packet.Modifier)
    assert.Equal(t, float32(0.1), packet.Sampling)
```